vscode/test/suites/extensionUtils.ts

```diff
@@ -31,3 +31,66 @@ export async function activateExtension() {
     `qsharp-tests: activate() completed in ${performance.now() - start}ms`,
   );
 }
+
+/**
+ * Waits until a condition is met, with a timeout.
+ *
+ * @param condition A function that returns true when the condition is met
+ * @param wakeUpOn The event that we want to wake up to check the condition
+ * @param timeoutMs If the condition is not met by this timeout, this function will throw
+ * @param timeoutErrorMsg The custom error message to throw if the condition is not met
+ */
+export async function waitForCondition(
+  condition: (...args: any[]) => boolean,
+  wakeUpOn: vscode.Event<any>,
+  timeoutMs: number,
+  timeoutErrorMsg: string,
+) {
+  let disposable: vscode.Disposable | undefined;
+  await new Promise<void>((resolve, reject) => {
+    let done = false;
+    setTimeout(() => {
+      if (!done) {
+        reject(new Error(timeoutErrorMsg));
+      }
+    }, timeoutMs);
+
+    disposable = wakeUpOn(() => {
+      if (!done && condition()) {
+        done = true;
+        resolve();
+      }
+    });
+
+    // Resolve immediately if condition is already met
+    if (condition()) {
+      done = true;
+      resolve();
+    }
+  });
+  disposable?.dispose();
+}
+
+/**
+ * Convenience method to add a time delay.
+ *
+ * @deprecated While this function is convenient for local development,
+ * please do NOT use it in actual tests. Instead, use `waitForCondition`
+ * with an appropriate condition and wake up event.
+ *
+ * Adding an unconditional delay to tests is guaranteed to slow tests down
+ * unnecessarily. It also reduces reliability when we're not clear about
+ * exactly what we're waiting on.
+ */
+export async function delay(timeoutMs: number) {
+  try {
+    await waitForCondition(
+      () => false,
+      () => ({ dispose() {} }),
```

Comment:
What 'dispose' is this calling?

Reply:
This is really hard to follow, but I'm actually defining an object here with a `dispose()` method that does nothing. (The horribly ambiguous `() => { foo }`/`() => ({ foo })`  syntax!)

This is to satisfy the type of this parameter, `vscode.Event<any>`. `vscode.Event<any>` has to return a `Disposable` when invoked.

Reply:
Ah! Got it.

I think part of what confused me is that in waitForCondition, disposable is `vscode.Disposable | undefined;` and its checked for nullness before invoking `disposable?.dispose();`, so seems like you shouldn't have to provide it if it's optional.

Reply:
Maybe the 'disposable' above shouldn't be nullable if invoking an Event<> is required to return one?

Reply:
Took another look, and I don't think I can change it. Even though in practice `disposable` should always be set, the null-check is necessary to make TypeScript happy. `disposable` needs to be declared outside the promise (so we can dispose after the promise resolves) but it's initialized _inside_ the promise, so we have to assume it's nullable in the outer scope.